Support directives and bindings on dynamically-created components

crisbeto opened the pull request:
These changes expand the APIs for creating components dynamically (`createComponent`, `ViewContainerRef.createComponent` and `ComponentFactory.create`) to allow users to apply directives on the component, as well as to bind to inputs/outputs in a targeted manner either on the component or one of the directives, for example:

```ts
import {createComponent, signal, inputBinding, outputBinding} from '@angular/core';

const canClose = signal(false);

// Create MyDialog
createComponent(MyDialog, {
  bindings: [
    // Bind a signal to the `canClose` input.
    inputBinding('canClose', canClose),

    // Listen for the `onClose` event specifically on the dialog.
    outputBinding<Result>('onClose', result => console.log(result)),
  ],
  directives: [
    // Apply the `FocusTrap` directive to `MyDialog` without any bindings.
    FocusTrap,

    // Apply the `HasColor` directive to `MyDialog` and bind the `red` value to its `color` input.
    // The callback to `inputBinding` is invoked on each change detection.
    {
      type: HasColor,
      bindings: [inputBinding('color', () => 'red')]
    }
  ]
});
```

Note that while it has been possible to do some these things through other APIs in the past (e.g. `setInput`), these new APIs have a few advantages:
1. They go through the same mechanisms as the template, meaning that they'll behave in the same way as input and output bindings in the template. For example, `outputBinding` will clean up automatically when the component is destroyed.
2. They are able to target specific directives, whereas existing APIs apply to all directives at the same time which might not be desirable in some cases.
3. The new APIs are tree-shakeable so their code won't make it into the bundle unless you're actually using it, whereas existing APIs like `setInput` are defined as method that cannot be deleted.
4. It provides us with a unified way of defining bindings in TypeScript that can be rolled out in other places like `TestBed` and `hostBindings`.

crisbeto committed:
refactor(core): move duplicate directive check outside of host directives

The check that verifies that there are no duplicates in the directives array was only running after host directive matching since that was the only case when it can happen. After the upcoming changes that won't be the case anymore so these changes move it always run after directive matching.

I also did some additional cleanup by adding comments and by not lazily initializing the `allDirectiveDefs` array when matching host directives. The array is guaranteed to be defined since earlier in the function we verify that there's at least one def with host directives.

crisbeto committed:
feat(core): add the ability to apply directives to dynamically-created components

Updates `createComponent`, `ViewContainerRef.createComponent` and `ComponentFactory.create` to allow the user to specify directives that should be applied when creating the component.

crisbeto committed:
refactor(core): move createComponent tests into a separate file

Moves the tests for `createComponent` into their own file since the `component_spec.ts` was a bit too generic and was accumulating all sorts of tests.

crisbeto committed:
refactor(core): set up underlying input binding symbols

Sets up the symbols used to power the upcoming `inputBinding` functionality.

I also fixed that `setDirectiveInput` was incorrectly only allowing strings for the `value` parameter.

crisbeto committed:
refactor(core): add assertion parameter to getDirectiveDef

Some upcoming functionality won't work if we can't retrieve a directive definition from a class. These changes add a `throwIfNotFound` to `getDirectiveDef`, similar to `getNgModuleDef`, to avoid duplication in such cases.

crisbeto committed:
fix(core): input targeting not checking if input exists on host

Fixes that the `setDirectiveInput` function wasn't checkin if an input exists before writing to it which can lead to assertion errors.

crisbeto committed:
feat(core): add input binding support to dynamically-created components

Adds the ability to bind to inputs on dynamically-created components, either by targeting the component itself or one of its directives. The new API looks as follows:

```ts
const value = signal(123);

createComponent(MyComp, {
  // Bind the value `'hello'` to `someInput` of `MyComp`.
  bindings: [inputBinding('someInput', () => 'hello')],

  directives: [{
    type: MyDir,
    // Bind the `value` signal to the `otherInput` of `MyDir`.
    bindings: [inputBinding('otherInput', value)]
  }]
});
```

This behavior overlaps with `ComponentRef.setInput`, with a few key differences:
1. `setInput` sets the value on *all* inputs whereas `inputBinding` only targets the specified directive and its host directives. This makes it easier to know which directive you're targeting.
2. `inputBinding` is executed as if it's in a template, making it consistent with how bindings behave for selector-matched components, whereas `setInput` executes outside the lifecycle of the component.
3. It resolves a long-standing issue with `setInput` where it wasn't possible to set the initial value of an input before the first change detection run.

Currently `inputBinding` is used only for `createComponent`, `ViewContainerRef.createComponent` and `ComponentFactory.create`, however it is going to be base for more APIs in the future.

crisbeto committed:
fix(core): do not allow setInput to be used with inputBinding

Calling `setInput` while the component already has an `inputBinding` active can lead to inconsistent state. These changes add an error that will be thrown if that's the case.

eneajaho commented:
Also it brings us one step closer to https://github.com/angular/angular/issues/43120,  I guess.

crisbeto committed:
feat(core): support listening to outputs on dynamically-created components

Adds the new `outputBinding` function that allows users to listen to outputs on dynamically-created components in a similar way to templates. For example, here we create an instance of `MyCheckbox` and listen to its `onChange` event:

```ts
interface CheckboxChange {
  value: string;
}

createComponent(MyCheckbox, {
  bindings: [
   outputBinding<CheckboxChange>('onChange', event => console.log(event.value))
  ],
});
```

Note that while it has always been possible to listen to events like this by getting a hold of of the instance and subscribing to it, there are a few key differences:
1. `outputBinding` behaves in the same way as if the event was bound in a template which comes with some behaviors like forwarding errors to the `ErrorHandler` and marking the view as dirty.
2. With `outputBinding` the listeners will be cleaned up automatically when the component is destroyed.
3. `outputBinding` accounts for host directive outputs by binding to them through the host. E.g. if the `onChange` event above was coming from a host directive, `outputBinding` would bind to it automatically.

Currently `outputBinding` is available only in `createComponent`, `ViewContainerRef.createComponent` and `ComponentFactory.create`, but it will serve as a base for APIs in the future.

crisbeto commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -1,7 +1,7 @@
   "cli-hello-world": {
     "uncompressed": {
-      "main": 132425,
+      "main": 137893,
```
These changes definitely added some bytes, but I doubt it was 5kb so I'm not totally sure where it comes from.

crisbeto commented on packages/core/src/render3/instructions/listener.ts:
```diff
@@ -378,3 +378,80 @@ function isOutputSubscribable(value: unknown): value is SubscribableOutput<unkno
 }
+
+/** Listens to an output on a specific directive. */
+export function listenToDirectiveOutput(
```
I wasn't totally sure where to place this one, but it's basically a targeted version of the listener instruction.

sp90 commented:
Would this also help solve the #51878 or are they unrelated

crisbeto commented:
It's a step towards resolving https://github.com/angular/angular/issues/51878, but the main issue is that there's no way we can infer the type if an input is aliased.

osnoser1 commented:
What about making content projection and keeping the hydration strategy turned on? 🥲

Harpush commented:
Will the new API allow (later probably) host directives input/output binding from hosting component without exposing them?

demike commented:
Is it possible to bind to multiple directives and the component with one inputBinding? Or is this a scenario where ComponentRef.setInput should be used?

crisbeto commented:
@Harpush regarding host directives: allowing bindings directly to the host directives is something we're considering, but it will likely still be only to the exposed inputs/outputs.

Also yes, we're considering adding `twoWayBinding` and potentially content projection as bindings as well.


@demike regarding targeting: we intentionally allow only targeted writes since the intent is much clearer compared to `setInput`. That being said, the `inputBinding` doesn't have any special compiler magic so you can pass it around if you want to reuse it, e.g.

```
const disabledBinding = inputBinding('disabled', () => true);

createComponent(Comp, {
  directives: [{type: One, bindings: [disabled, ...otherBindings]}, {type: Two, bindings: [disabled]}]
});
```

wszgrcy commented:
I have also implemented a method for dynamic instruction loading
Because I haven't found a job, I haven't made it public either
But currently, someone has proposed PR. I have also made it public for everyone's reference
https://github.com/wszgrcy/dynamic-component-define

Jordan-Hall commented:
Surely any change like this needs a RFC. 


Questions: Wouldn't createComponent and cresteDirective etc be a good replacement for new authoring as a way of removing decorators?

If authoring goes functional is they actually going to be a real need for this? 

Is this just going to be a stop gap and if so doesn't that make end migration harder?

demike commented:
> @Harpush regarding host directives: allowing bindings directly to the host directives is something we're considering, but it will likely still be only to the exposed inputs/outputs.
> 
> Also yes, we're considering adding `twoWayBinding` and potentially content projection as bindings as well.
> 
> @demike regarding targeting: we intentionally allow only targeted writes since the intent is much clearer compared to `setInput`. That being said, the `inputBinding` doesn't have any special compiler magic so you can pass it around if you want to reuse it, e.g.
> 
> ```
> const disabledBinding = inputBinding('disabled', () => true);
> 
> createComponent(Comp, {
>   directives: [{type: One, bindings: [disabled, ...otherBindings]}, {type: Two, bindings: [disabled]}]
> });
> ```

So this approach assumes that all inputs are known.
**Is it possible to ignore missing bindings like inputBinding('disable', () => true, { ignoreMissingInput: true }**

Scenario json based templating mechanism
```ts
const template = {
  componenId: "mycomponent", // dumb component --> id is used for lookup in a registry to get the class for createComponent
  someComponentInput: true,
  bind: "some.resource.identifier", // input for directive binding (like ngModel but for websocket data
  children: [
   // ...
  ]
  // ...
}
```
In the template engine the bindTo is well known --> create the directive for that and set the input
but the component that gets instantiated might also be intereseted in the binding address (`bind`) . i.e.: if it is a complex component and wants to bind to sub resources.

Therefore I would pass all members of the template to inputBinding(inputName, inputValue, { ignoreMissingInput: true }) on the component
and `template.bind` as input binding to the well known binding directive

GuillaumeREIGNAT commented:
I await this change with enthusiasm! 

Will it be possible to apply defer/hydrate on these components through createComponent? Currently, I'm struggling to find a way to do that.

pkozlowski-opensource commented on packages/core/test/acceptance/create_component_spec.ts:
```diff
@@ -0,0 +1,1324 @@
+    },
+  );
+
+  describe('attaching directives to root component', () => {
```
nit: this test file becomes a bit long, I could see how thopse new tests are moved to a separate file (something like `create_component_with_directives_spec.ts`

pkozlowski-opensource commented on packages/core/test/acceptance/create_component_spec.ts:
```diff
@@ -0,0 +1,1324 @@
+      }
+
+      @Component({template: ''})
+      class HostComponent {}
```
Not sure if you've got a test like this later on but it would be beneficial to have one where there is a directive and a host component with the same input and we verify that `setInput` writes to both. This should work until we decide to do something with `setInput`

pkozlowski-opensource commented on packages/core/test/acceptance/create_component_spec.ts:
```diff
@@ -0,0 +1,1324 @@
+          directives: [NotADir],
+        });
+      }).toThrowError(/Type NotADir does not have 'ɵdir' property/);
+    });
```
Do we make any assumption about the attached directives being standalone? I guess we should? Plus add a test that it throws when a non-standalone directive is attached.

pkozlowski-opensource commented on packages/core/test/acceptance/create_component_spec.ts:
```diff
@@ -0,0 +1,1324 @@
+          bindings: [outputBinding('click', () => {})],
+        });
+      }).toThrowError(/RootComp does not have an output with a public name of "click"/);
+    });
```
Q: the current 'listener' instruction would subscribe to an output _and_ add a DOM event listener. I guess this is _not_ the logic we want here. We should add a test to verify.

pkozlowski-opensource commented on goldens/public-api/core/index.api.md:
```diff
@@ -192,6 +192,18 @@ export interface BaseResourceOptions<T, R> {
 }
 
+// @public
+export interface Binding {
```
Should this interface end up in the public API space? It looks like it is something more of the framework-internal?

pkozlowski-opensource replied:
The other approach we could take is to have a public and private side of the API. The public API would be almost like an opaque token (unless there are operations we expect people to invoke on those) and the private / internal one would extend it. 

I would very much like to avoid the situation we have with `QueryList` where the public API mixes both functionality meant for the end user and the one to be invoked internally by the framework.

crisbeto replied:
Dropped it from the public API for now and will monitor if the example I mentioned above comes up in issues.

crisbeto committed:
refactor(core): allow InputBinding and OutputBinding to be tree shaken

Reworks the `InputBinding` and `OutputBinding` functionality to be in object literals constructed in functions, rather than classes, because it seems like Terser was having a hard time tree shaking the classes when the functions weren't used.

thePunderWoman reviewed (approved):
reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

mmalerba commented:
This PR was merged into the repository by commit a3cb7b94978654f5bd89254bea49685ac18efdbc.

The changes were merged into the following branches: main

kukjevov commented:
> This PR also fixes this issue #47728 (even though this one was closed) 🎉

Awesome, love it.
